devices/src/virtio/vsock/unix/muxer_rxq.rs

```diff
@@ -0,0 +1,139 @@
+/// This walk is performed here, as part of building an RX queue from the connection pool. When an
+/// out-of-sync is drained, the muxer will discard it, and attempt to rebuild a synced one.
+///
+use std::collections::{HashMap, VecDeque};
+
+use super::super::VsockChannel;
+use super::defs;
+use super::muxer::{ConnMapKey, MuxerRx};
+use super::MuxerConnection;
+use super::{Error, Result};
+
+/// The muxer RX queue.
+///
+pub struct MuxerRxQ {
+    /// The RX queue data.
+    q: VecDeque<MuxerRx>,
+    /// The RX queue sync status.
+    synced: bool,
+}
+
+impl MuxerRxQ {
+    const SIZE: usize = defs::MUXER_RXQ_SIZE;
+
+    /// Trivial RX queue constructor.
+    ///
+    pub fn new() -> Self {
+        Self {
+            q: VecDeque::with_capacity(Self::SIZE),
+            synced: true,
+        }
+    }
+
+    /// Attempt to build an RX queue, that is synchronized to the connection pool.
+    /// Note: the resulting queue may still be desynchronized, if there are too many connections
+    ///       that have pending RX data. In that case, the muxer will first drain this queue, and
+    ///       then try again to build a synchronized one.
+    ///
+    pub fn from_conn_map(conn_map: &HashMap<ConnMapKey, MuxerConnection>) -> Self {
+        let mut q = VecDeque::new();
+        let mut synced = true;
+
+        for (key, conn) in conn_map.iter() {
+            if !conn.has_pending_rx() {
+                continue;
+            }
+            if q.len() >= Self::SIZE {
+                synced = false;
+                break;
+            }
+            q.push_back(MuxerRx::ConnRx(*key));
+        }
+        Self { q, synced }
+    }
+
+    /// Push a new RX item to the queue.
+    ///
+    /// A push will fail when:
+    /// - trying to push a connection key onto an out-of-sync, or full queue; or
+    /// - trying to push an RST onto a queue already full of RSTs.
+    /// RSTs take precedence over connections, because connections can always be queried for pending
+    /// RX data later. Aside from this queue, there is no other storage for RSTs, so, failing to
+    /// push one means that we have to drop the packet.
+    ///
+    /// Returns:
+    /// - `Ok(())` if the new item has been successfully queued; or
+    /// - `Err(Error::QueueFull)` if there was no room left in the queue.
+    ///
+    pub fn push(&mut self, rx: MuxerRx) -> Result<()> {
+        // Pushing to a non-full, synchronized queue will always succeed.
+        if self.is_synced() && !self.is_full() {
+            self.q.push_back(rx);
+            return Ok(());
+        }
+
+        match rx {
+            MuxerRx::RstPkt { .. } => {
+                // If we just failed to push an RST packet, we'll look through the queue, trying to
+                // find a connection key that we could evict. This way, the queue does lose sync, but
+                // we don't drop any packets.
+                for qi in self.q.iter_mut().rev() {
```

Comment:
This is just walking the queue backwards. `rev()` applies to the iterator only, not the actual data. I'm searching backwards in an effort to maintain packet delivery priority. When a connection  is evicted, it will have to wait until the queue is drained, before it gets asked for data again. So, it makes sense to evict the more recently added connections, so as to try to avoid delaying the first comers.